// socialblade/dist/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const matrix_1 = require("./matrix");
const platform_1 = require("./platform");
class SocialBlade {
    constructor(client_id, access_token, options = {}) {
        this.api = new matrix_1.Matrix(client_id, access_token, options);
        this.instagram = new platform_1.PlatformWithTop(this.api, 'instagram', 'followers');
        this.youtube = new platform_1.PlatformWithTop(this.api, 'youtube', 'subscribers');
        this.facebook = new platform_1.PlatformWithTop(this.api, 'facebook', 'likes');
        this.twitter = new platform_1.PlatformWithTop(this.api, 'twitter', 'followers');
        this.twitch = new platform_1.PlatformWithTop(this.api, 'twitch', 'followers');
        this.tiktok = new platform_1.PlatformWithTop(this.api, 'tiktok', 'followers');
        this.dailymotion = new platform_1.Platform(this.api, 'dailymotion');
        this.storyfire = new platform_1.Platform(this.api, 'storyfire');
        this.dlive = new platform_1.Platform(this.api, 'dlive');
        this.trovo = new platform_1.Platform(this.api, 'trovo');
    }
}
exports.default = SocialBlade;
